tensorflow/core/platform/default/logging.cc

```diff
@@ -31,28 +31,41 @@ limitations under the License.
 #include <string.h>
 #include <time.h>
 
-#include <string>
+#include <algorithm>
+#include <mutex>
+#include <queue>
 #include <unordered_map>
+#include <vector>
 
 namespace tensorflow {
+namespace internal {
 
-void TFAddLogSink(TFLogSink* sink) {
-  // LogSink is not implemented.
-  // If necessary, one can add the log sink support as follows.
-  // 1. Define a global vector<TFLogSink> to keep track of all registered
-  //    TFLogSink objects. Protect the global vector with mutex to make it
-  //    thread-safe.
-  // 2. Add/remove elements from the global vector<TFLogSink> in TFAddLogSink
-  //    and TFRemoveLogSink function
-  // 3. Add logic in LogMessage::GenerateLogMessage() below to dispatch log
-  //    messages to all the registered log sinks.
-}
+class TFLogSinks {
+  public:
+    static TFLogSinks& Instance();
 
-void TFRemoveLogSink(TFLogSink* sink) {
-  // LogSink is not implemented.
-}
+    void Add(TFLogSink* sink);
+    void Remove(TFLogSink* sink);
+    void Send(const TFLogEntry& entry);
 
-namespace internal {
+  private:
+    TFLogSinks() = default;
+    std::vector<std::unique_ptr<TFLogSink>>::iterator FindSink(TFLogSink* sink);
+    void SendToSink(TFLogSink& sink, const TFLogEntry& entry);
+
+#ifndef NO_DEFAULT_LOGGER
+    TFDefaultLogSink default_;
+#else
+    std::queue<TFLogEntry> queue_;
```

Comment:
This is like `int integer;`.  Let's use more descriptive name e.g., `log_entry_buffer_;`

Reply:
I thought the fact that it's in this class is descriptive-enough, but fair enough. It's not a buffer, it's a queue, so I'll specify it as `log_entry_queue_`.